jmespath/functions.py

```diff
@@ -35,6 +35,10 @@
 }
 
 
+def is_listlike(arg):
+        return hasattr(arg, "sort")
```

Comment:
I think we should be more careful here. This is a *very* wide door you're opening, since many, many things can have a `sort` method. I'd recommend being more cautious and use something more aligned with the original code:
```suggestion
def is_listlike(arg):
        return isinstance(arg, list) or is_ndarray(arg)
```

Reply:
Opening this to unknown listlike containers is the intention here, the door in python is already wide open as runtime type checks are much more of the exception than the norm in pythonic code (the implementation of `is_ndarray` would have the same problem as `__finalize_array__` may even be removed/changed silently in certain version of numpy). If the associated deserializer can return custom/unknown types (like `pickle.dumps` or `eval()`) this library either way provides no security layer as one can simply override `list` and it will pass the stricter `isinstance(value, list)` check and execute any serialized code.

This current implementation additionally works for:
* collections.UserList

but unfortunately not:
* tuple (a.k.a immutable list)
* xarray.DataArray
* pandas.Series

In most cases JMESPath.py would only ever exposed to the simple python containers returned from `json.dumps`, but in our case this now includes `msgpack_numpy`, `ast.literal_eval` and potentially other deserializers.

Looking at this again a better implementation to support those is to check for  `__getitem__` and `count` where `__getitem__` is the primary important interface for an indexable container and `count` is what's common to all of the above and excludes `dict`, `set`:

```python
set(dir(list)).intersection(dir(tuple)).intersection(set(dir(xarray.DataArray))).intersection(set(dir(pandas.Series))) - set(dir(dict)) - set(dir(set))
>> {'__mul__', '__add__', '__rmul__', 'count'}
```

something like:

```python
def is_listlike(arg):
    return hasattr(arg, "__getitem__") and (
        hasattr(arg, "sort")
        or (hasattr(arg, "count") and not isinstance(arg, (str, bytes)))
    )
```

Reply:
OK, I think I misunderstood the full intent of the MR. I was focusing on numpy solely, while the changes address two different issues: first, they add support for any list-like type, and they additionally also make sure that ndarrays in particular work well with some further fine tuning of the equality comparison.

It might have been just me, but I think this double purpose could be better reflected to avoid further confusions when trying to submit this upstream. I think part of my confusion came from the title, which starts with "Numpy", drawing your attention to that aspect of the changes. But also because the only new tests present here are around numpy arrays, even though in principle, as you mention, this should work with other types.

I need to jump into a meeting now, but will continue with this after that -- I think the newer `is_listlike` function looks a bit better, but it still looks a bit complex. Also the language of "listlike" might not be what we want -- maybe we should switch to using `is_sequence`? That's the more usual, generic way of referring to containers that can be randomly accessed by index.

Reply:
So this is where the complicated part starts, `numpy.ndarray` does not implement the [Sequence](https://docs.python.org/3/library/stdtypes.html#sequence-types-list-tuple-range) interface (neither does pandas.Series or xarray.DataArray, and the reason for creating a new "listlike" term)

Reply:
That looks prettttty nice now :). I also like the usage of `__array__` for numpy-like array discrimination, and of `isinstance(arg, Sequence)` too for getting lists and tuples. All in all, much, much nicer :)